### PR TITLE
fix(nemesis.py): restore only OSS snapshots on OSS

### PR DESCRIPTION
**Pull request by fgelcer**
since enterprise has an in-memory feature that
is not avaiable on OSS, we got a restore failing
on `malformed_sstable_exception`:
```
Column in_memory missing in current schema
```

Fixes: scylladb/scylladb#15846

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by fgelcer** [comment]
[testing here](https://jenkins.scylladb.com/job/scylla-staging/job/fabio/job/manager/job/restore-nemesis/2/)

File: sdcm/nemesis.py [hunk 1]
```diff
@@ -2774,7 +2774,15 @@ def choose_snapshot(snapshots_dict):
 
             self.use_nemesis_seed()
             chosen_snapshot_size = random.choice(fitting_snapshot_sizes)
-            snapshot_tag = random.choice(list(snapshot_groups_by_size[chosen_snapshot_size]["snapshots"].keys()))
+            if self.cluster.nodes[0].is_enterprise:
+                snapshot_tag = random.choice(list(snapshot_groups_by_size[chosen_snapshot_size]["snapshots"].keys()))
+            else:
+                oss_regex = r'\d\.\d\.\d'
```
Comment (bhalevy) [inline sdcm/nemesis.py:2780]: This seems fragile as a way to detect OSS versions

In any case, can't we have a double digit patch version?

Comment (bhalevy) [reply sdcm/nemesis.py:2780]: Also the version in the master branch would have `.dev` in it.
I'm not sure if this regex will match it.

Comment (fgelcer) [reply sdcm/nemesis.py:2780]: first of all, this is related to the already backed up snapshot versions, it has nothing to do with system under tests..
what does test if this is enterprise or OSS is `self.cluster.nodes[0].is_enterprise`

anyway, i thought of a better way to represent it, and will do the changes, and push it soon

Comment (fruch) [reply sdcm/nemesis.py:2780]: We have better functions to handle this kind of logic, and also we have our own implementation of versions comparing exact for matching those

Anyhow in memory is remove, and 2023.1 doesn't have it's so it should work if we have snapshot from 2023.1

Comment (fgelcer) [reply sdcm/nemesis.py]: > We have better functions to handle this kind of logic, and also we have our own implementation of versions comparing exact for matching those
> 
> Anyhow in memory is remove, and 2023.1 doesn't have it's so it should work if we have snapshot from 2023.1

[these are the snapshots](https://github.com/scylladb/scylla-cluster-tests/blob/master/defaults/manager_persistent_snapshots.yaml) we have... it means that, snapshots from `2022.1.X` and `2022.2.Y` are not usable, not even on enterprise master branch?
if that is the case, a much bigger refactor must be done in here

Comment (fgelcer) [reply sdcm/nemesis.py]: @bhalevy ^^

Comment (bhalevy) [reply sdcm/nemesis.py]: > Anyhow in memory is remove, and 2023.1 doesn't have it's so it should work if we have snapshot from 2023.1

@fruch it wasn't be removed from the schema tables, otherwise upgrade from 2022.X to 2023.1 would not work.

Comment (fruch) [reply sdcm/nemesis.py]: > > Anyhow in memory is remove, and 2023.1 doesn't have it's so it should work if we have snapshot from 2023.1
> 
> @fruch it wasn't be removed from the schema tables, otherwise upgrade from 2022.X to 2023.1 would not work.

So it's gonna stay there for good ?

Comment (bhalevy) [reply sdcm/nemesis.py]: We are not actively working on removing it.
I think the best way would be to have the manager restore the schema using some high-level logic.
There are other schema-related incompatibilities between enterprise and open source like IncrementalCompactionStrategy, encryption at rest, and more, that need to be dealt with.

Comment (fgelcer) [reply sdcm/nemesis.py]: changed the logic, running it with latest code [in here](https://jenkins.scylladb.com/job/scylla-staging/job/fabio/job/manager/job/restore-nemesis/4/)

File: sdcm/nemesis.py [hunk 2]
```diff
@@ -2811,7 +2818,8 @@ def execute_data_validation_thread(command_template, keyspace_name, number_of_ro
             self.log.info("Restoring the schema of the keyspace '%s'", chosen_snapshot_info["keyspace_name"])
             restore_task = mgr_cluster.create_restore_task(restore_schema=True, location_list=location_list,
                                                            snapshot_tag=chosen_snapshot_tag)
-            restore_task.wait_and_get_final_status(step=10, timeout=120)
+            step = 10
+            restore_task.wait_and_get_final_status(step=step, timeout=chosen_snapshot_info['expected_timeout'] / step)
```
Comment (bhalevy) [inline sdcm/nemesis.py:2822]: This hunk looks unrelated

Comment (fgelcer) [reply sdcm/nemesis.py]: well, it is related to the fact that the test is still not passing, and now it is failing on this call, timing out while waiting for this task to end... but now i have moved it to its own commit

**Review by bhalevy — APPROVED** [review]
lgtm

**Review by fruch — APPROVED** [review]
LGTM

beside the fact this one closes:
https://github.com/scylladb/scylladb/issues/15846

it's a workaround, and this is still an issue between scylla-core and scylla-manager

**Comment by fgelcer** [comment]
@fruch , the part of the choosing the right snapshot to restore seems now fine, but it is failing with timeout.

maybe i don't fully understand how the implementation works, but `step` is the amount of cycles we will wait for `timeout`, right? if that is the case, there is something wrong with the call, since it fails with timing out after 90 seconds (while the timeout is 900)

i'm still trying to debug it, but i would be happy to get some directions, if you know

**Comment by fgelcer** [comment]
> LGTM
> 
> beside the fact this one closes: [scylladb/scylladb#15846](https://github.com/scylladb/scylladb/issues/15846)
> 
> it's a workaround, and this is still an issue between scylla-core and scylla-manager

if you think that htis is still an issue, then i can remove the `Fixes` and use `Refs` instead

**Comment by fruch** [comment]
> > LGTM
> > beside the fact this one closes: [scylladb/scylladb#15846](https://github.com/scylladb/scylladb/issues/15846)
> > it's a workaround, and this is still an issue between scylla-core and scylla-manager
> 
> if you think that htis is still an issue, then i can remove the `Fixes` and use `Refs` instead

nevermind, it's merged, and I've reopen the issue

**Comment by fruch** [comment]
> @fruch , the part of the choosing the right snapshot to restore seems now fine, but it is failing with timeout.
> 
> maybe i don't fully understand how the implementation works, but `step` is the amount of cycles we will wait for `timeout`, right? if that is the case, there is something wrong with the call, since it fails with timing out after 90 seconds (while the timeout is 900)
> 
> i'm still trying to debug it, but i would be happy to get some directions, if you know

docs say this:

```
def wait_for(func, step=1, text=None, timeout=None, throw_exc=True, stop_event=None, **kwargs):  # pylint: disable=too-many-arguments
    """
    Wrapper function to wait with timeout option.

    :param func: Function to evaluate.
    :param step: Time to sleep between attempts in seconds
    :param text: Text to print while waiting, for debug purposes
    :param timeout: Timeout in seconds
    :param throw_exc: Raise exception if timeout expired, but disrupt_func result is not True
    :param kwargs: Keyword arguments to disrupt_func
    :param stop_event: instance of threading.Event class to stop retrying
    :return: Return value of disrupt_func.
    """
```

so step it time we sleep between attempts, and timeout is the whole timeout.
regardless you are changing the time it waits for schema restore, it shouldn't be extended
if it's timing out, we have a bug needs to be handled

are you testing with master or with 5.4 ?

**Comment by fgelcer** [comment]
> > @fruch , the part of the choosing the right snapshot to restore seems now fine, but it is failing with timeout.
> > maybe i don't fully understand how the implementation works, but `step` is the amount of cycles we will wait for `timeout`, right? if that is the case, there is something wrong with the call, since it fails with timing out after 90 seconds (while the timeout is 900)
> > i'm still trying to debug it, but i would be happy to get some directions, if you know
> 
> docs say this:
> 
> ```python
> def wait_for(func, step=1, text=None, timeout=None, throw_exc=True, stop_event=None, **kwargs):  # pylint: disable=too-many-arguments
>     """
>     Wrapper function to wait with timeout option.
> 
>     :param func: Function to evaluate.
>     :param step: Time to sleep between attempts in seconds
>     :param text: Text to print while waiting, for debug purposes
>     :param timeout: Timeout in seconds
>     :param throw_exc: Raise exception if timeout expired, but disrupt_func result is not True
>     :param kwargs: Keyword arguments to disrupt_func
>     :param stop_event: instance of threading.Event class to stop retrying
>     :return: Return value of disrupt_func.
>     """
> ```
> 
> so step it time we sleep between attempts, and timeout is the whole timeout. regardless you are changing the time it waits for schema restore, it shouldn't be extended if it's timing out, we have a bug needs to be handled
> 
> are you testing with master or with 5.4 ?

yes, aligning it in #6737

tests failing (that led me to this fix) was in master, but same applies to 5.4
